protocol/msgpack/base_iterator.go

```diff
@@ -61,15 +66,40 @@ func (it *baseIterator) err() error       { return it.decodeErr }
 func (it *baseIterator) setErr(err error) { it.decodeErr = err }
 
 func (it *baseIterator) decodePolicy() policy.Policy {
-	numExpectedFields, numActualFields, ok := it.checkNumFieldsForType(policyType)
+	numActualFields := it.decodeNumObjectFields()
+	policyType := it.decodeObjectType()
+
+	numExpectedFields, numActualFields, ok := it.checkNumFieldsForTypeWithActual(
+		policyType,
+		numActualFields,
+	)
 	if !ok {
 		return policy.EmptyPolicy
 	}
-	resolution := it.decodeResolution()
-	retention := it.decodeRetention()
-	p := policy.NewPolicy(resolution.Window, resolution.Precision, time.Duration(retention))
-	it.skip(numActualFields - numExpectedFields)
-	return p
+
+	switch policyType {
+	case rawPolicyType:
+		resolution := it.decodeResolution()
+		retention := it.decodeRetention()
+		p := policy.NewPolicy(resolution.Window, resolution.Precision, time.Duration(retention))
+		it.skip(numActualFields - numExpectedFields)
+		return p
+	case compressedPolicyType:
+		id := it.decodeVarint()
+		if it.decodeErr != nil {
+			return policy.EmptyPolicy
+		}
+		p, ok := it.policyDecompressor.Get(id)
+		if !ok {
+			it.decodeErr = fmt.Errorf("unrecognized compression policy id: %v", id)
+			return policy.EmptyPolicy
```

Comment:
If the iterator error is set, the caller should check for error and not apply the policy.